### PR TITLE
Add encounter model and link to BPs

### PR DESCRIPTION
**Pull request by sanchita**
Pivotol story -> https://www.pivotaltracker.com/story/show/168448997

FYI: This PR has two database schema files hence the high LOC.

### CODE REVIEW

File: app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt [hunk 1]
```diff
@@ -3329,6 +3329,29 @@ class DatabaseMigrationAndroidTest {
     val dbV47 = helper.migrateTo(47)
     dbV47.assertTableDoesNotExist("HomescreenIllustration")
   }
+
+  @Test
+  fun migration_from_47_to_48_should_create_encounter_table() {
+    //given
+    val tableName = "Encounter"
```
Comment (ragunathjawahar) [inline app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt]: Call this variable `encounterTable` instead? Lowers cognitive overhead.

File: app/src/main/java/org/simple/clinic/storage/Migration_47_48.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    CREATE TABLE IF NOT EXISTS "Encounter" (
+    "uuid" TEXT NOT NULL,
+    "patientUuid" TEXT NOT NULL,
+    "encounteredOn" TEXT NOT NULL,
+    "createdAt" TEXT NOT NULL,
+    "updatedAt" TEXT NOT NULL,
+    "deletedAt" TEXT,
+    PRIMARY KEY("uuid")
+    )
+    """)
```
Comment (ragunathjawahar) [inline app/src/main/java/org/simple/clinic/storage/Migration_47_48.kt]: Please check formatting?

File: app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt [hunk 3]
```diff
@@ -3352,6 +3355,85 @@ class DatabaseMigrationAndroidTest {
         "deletedAt")
     )
   }
+
+  @Test
+  fun migration_from_48_to_49_create_only_one_encounter_for_per_day_but_add_encounterUuid_to_all_BPs() {
+    //given
+    val tableName = "BloodPressureMeasurement"
```
Comment (ragunathjawahar) [inline app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt]: variable name `bloodPressureMeasurementTable`?

File: app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt [hunk 4]
```diff
@@ -3352,6 +3355,85 @@ class DatabaseMigrationAndroidTest {
...
+
+    val uuidForFirstBP = "3eb96138-5270-419c-bcdf-86db95d0bc8b"
+    val uuidForSecondBP = "a2455b0c-cf14-46c9-9f73-6e18972982b0"
+    val facilityUuid = "0086ba86-314c-49d9-8822-2563e53bed92"
+    val patientUuid = "c4233170-feb2-3d34-8115-98fef5a68837"
+    val recordedAtForFirstBP = "2018-07-03T00:00:00Z"
+    val recordedAtForSecondBP = "2018-07-03T01:30:00Z"
+    val createdAtForFirstBP = "2018-07-03T00:00:00Z"
+    val createdAtForSecondBP = "2018-07-03T01:30:00Z"
```
Comment (ragunathjawahar) [inline app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt]: Can we do `val createdAtForFirstBp = recordedAtForFirstBP`? and `recordedAtForSecondBP` as well?

Comment (sanchita) [reply app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt]: We can do that but we don't expect both these values to be the same always.
But I understand what you mean. I will change the `Instant` for both `createdAt`.

File: app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt [hunk 5]
```diff
@@ -3352,6 +3355,85 @@ class DatabaseMigrationAndroidTest {
...
+
+    val db_47 = helper.createDatabase(version = 48)
+    db_47.insert(tableName, mapOf(
+        "uuid" to uuidForFirstBP,
+        "systolic" to "120",
+        "diastolic" to "90",
+        "syncStatus" to "DONE",
+        "userUuid" to "d689285f-5c92-4735-80ef-c0e360ad3f1d",
```
Comment (ragunathjawahar) [inline app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt]: Extract this as a variable `"d689285f-5c92-4735-80ef-c0e360ad3f1d"` ?

File: app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt [hunk 6]
```diff
@@ -3352,6 +3355,85 @@ class DatabaseMigrationAndroidTest {
...
+        "deletedAt" to "null",
+        "recordedAt" to recordedAtForSecondBP
+    ))
+
+    //when
+    val db_49 = helper.migrateTo(version = 49)
+
+    //then
+    val expectedEncounteredOn = Instant.parse(recordedAtForFirstBP).toLocalDateAtZone(ZoneOffset.UTC).toString()
+    val expectedEncounterId = createUuid5(facilityUuid + patientUuid + expectedEncounteredOn).toString()
```
Comment (ragunathjawahar) [inline app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt]: `createUuid5(facilityUuid + patientUuid + expectedEncounteredOn)` this method signature reveals too much information and does not reveal intent, can we change the signature to `createUuid5(facilityUuid, patientUuid, expectedEncounteredOn)`?

File: app/src/androidTest/java/org/simple/clinic/TestData.kt [hunk 7]
```diff
@@ -676,7 +677,8 @@ class TestData @Inject constructor(
         createdAt = createdAt,
         updatedAt = updatedAt,
         deletedAt = deletedAt,
-        recordedAt = recordedAt
+        recordedAt = recordedAt,
+        encounterUuid = createUuid5(facilityUuid.toString() + patientUuid + recordedAt)
```
Comment (ragunathjawahar) [inline app/src/androidTest/java/org/simple/clinic/TestData.kt]: Same as mentioned above - function signature.

File: app/src/main/java/org/simple/clinic/bp/BloodPressureRepository.kt [hunk 8]
```diff
@@ -48,7 +49,8 @@ class BloodPressureRepository @Inject constructor(
                 createdAt = now,
                 updatedAt = now,
                 deletedAt = null,
-                recordedAt = recordedAt)
+                recordedAt = recordedAt,
+                encounterUuid = createUuid5(currentFacility.uuid.toString() + patientUuid + recordedAt))
```
Comment (ragunathjawahar) [inline app/src/main/java/org/simple/clinic/bp/BloodPressureRepository.kt]: Here and other usage sites as well.

File: app/src/main/java/org/simple/clinic/storage/Migration_48_49.kt [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+@Suppress("ClassName")
+class Migration_48_49 : Migration(48, 49) {
+
+  override fun migrate(database: SupportSQLiteDatabase) {
+    val tableName = "BloodPressureMeasurement"
+
+    database.inTransaction {
+      //Using a random UUID to give a default value to non-null encounterUuid temporarily.
+      //The encounterUuid will be replaced with the correct value when the query executes in next line.
+      database.execSQL(""" ALTER TABLE $tableName ADD COLUMN "encounterUuid" TEXT NOT NULL DEFAULT '05f9c798-1701-4379-b14a-b1b18d937a33'  """)
```
Comment (ragunathjawahar) [inline app/src/main/java/org/simple/clinic/storage/Migration_48_49.kt]: I don't understand why we are doing this - `05f9c798-1701-4379-b14a-b1b18d937a33` can you please explain?

File: app/src/test/java/org/simple/clinic/patient/PatientMocker.kt [hunk 10]
```diff
@@ -110,7 +111,9 @@ object PatientMocker {
         createdAt = createdAt,
         updatedAt = updatedAt,
         deletedAt = deletedAt,
-        recordedAt = recordedAt)
+        recordedAt = recordedAt,
+        encounterUuid = createUuid5(facilityUuid.toString() + patientUuid + recordedAt)
```
Comment (ragunathjawahar) [inline app/src/test/java/org/simple/clinic/patient/PatientMocker.kt]: +1 here too.

File: app/build.gradle [hunk 11]
```diff
@@ -307,6 +307,7 @@ dependencies {
     because "Traceur is incompatible with the version of RxJava this artifact pulls in transitively"
   }
   testImplementation "org.threeten:threetenbp:$versions.threeTenBp"
+  implementation "com.github.f4b6a3:uuid-creator:1.4.1"
```
Comment (ragunathjawahar) [inline app/build.gradle]: Would you consider using a library from a more reliable source?

https://github.com/cowtowncoder/java-uuid-generator

This library is from the Jackson folks.

File: app/src/main/java/org/simple/clinic/storage/Migration_47_48.kt (hunk 2 above)
Comment (sanchita) [reply app/src/main/java/org/simple/clinic/storage/Migration_47_48.kt]: The SQL query string doesn't get auto-formatted.
We should maybe add a formatter for these as well.

File: app/build.gradle (hunk 11 above)
Comment (sanchita) [reply app/build.gradle]: The library we are using right now is already tested with between app and server.
If you think we should still update, I will make this change.

File: app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt (hunk 6 above)
Comment (sanchita) [reply app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt]: I have added a separate method which is responsible for generating UUIDs for `Encounter`.
PTAL. :)

**Pull request merged**